README
Welcome to thelinkbox, a member of the CQiNet family.

http://CQiNet.sourceforge.net

Support: http://groups.yahoo.com/group/thelinkbox/

Thelinkbox is an voice over IP (VoIP) radio linking package for ham radio
operators that runs under FreeBSD and Linux, and hopefully other Posix 
compatible operating  systems.  It does not and probably never will run 
under Windows.

Thelinkbox is basically thebridge conferencing software with additional code 
to interface to radios. Although the primary purpose is for VoIP linking of a 
limited number of  repeaters "off grid", i.e. separate from the EchoLink and 
IRLP networks it can also be used on the Echolink or IRLP networks when 
properly configurated.  

Unlike some other VoIP systems thelinkbox also supports multiple radio 
ports. The ultimate goal is for thelinkbox to become a full featured
multiport hub or repeater controller as well as a VoIP application. I
believe the use of USB audio dongles under Linux will allow a modern
PC to support a larger number of radio ports than are likely to be
needed by even the largest repeater groups (> 16 ports), although I
have yet to test the limit in practice.


Here are some features:

1. Support for multiple simultaneous connections using different
codecs and protocols. Currently tlb supports ADPCM and GSM codecs and
the Echolink, RTP, and Speak freely protocols. When multiple sources
are active the audio is mixed.

2. The number of VoIP connections is limited only by system RAM and
available bandwidth.

3. The number of physical radio ports is limited only by available hardware, 
OS hardware support and CPU processing power.

4. All ports are completely independent: Each port may have a unique
set of DTMF commands. Each port can be connected to a distinct VoIP clients.

5. A full cross point matrix of connections between VoIP clients and
RF ports is supported.

6. Does not require the use of a central server or authentication
authority, all that is required is internet connectivity.

7. Support for either prerecorded PCM audio announcements or external
text to speech systems such as Festival.

8. CW and voice IDs.

9. Flexible general purpose telemetry tone generator that can be assigned
to various events and/or used by scripts.

10. "Permanent" connections, if a link is lost tlb will keep trying to
reconnect indefinitely.

11. Builtin software based DTMF and CTCSS encoders and decoders.

12. Open source software written in C and C++.

13. Support for inexpensive USB audio dongles including PTT, 
COS and CTCSS sense and frequency control.


#############################################
Updating from previous versions of thelinkbox
#############################################

New versions of thelinkbox are usually backwards compatible with configuration 
files from earlier releases HOWEVER new configuration file variables are added 
frequently.  Since the sample tlb.conf.sample file includes documentation on 
features which are not mentioned elsewhere it is worthwhile to review the 
sample configuration file with each release to discover new capabilities which 
you may want to take advantage of.  


###########
Portability
###########

A quick word on portability:  I've tried to the best of my ability and means
to make thelinkbox as portable as possible.  I've tested it under several
versions of FreeBSD and Linux, however *ALL* testing has been on the Intel x86
architecture.  

I've found that sound programming on Linux platforms can be very problematic.  
Results are very dependent on the sound board and its driver.  At least 50% of 
the sounds cards I've tested with had some issues that required special 
workarounds such as extra driver options.  Additionally for portability 
thelinkbox uses the OSS sound system API supported by many platforms, not the 
Linux specific ALSA sounds system.  Luckily the ALSA sound system includes an 
emulated OSS API for backwards compatibility.

At this time the USB subsystem is Linux specific.  I tried to make the USB code
portable by using libusb but I've run into a show stopper on FreeBSD, namely
if the USB audio device is claimed by the kernel's audio driver then libusb
is unable to access the device.


###################
Building on a *nix:
###################

Executive summary: (the usual)
        "tar xzf thelinkbox-{VERSION}.tgz"
        "cd thelinkbox-{VERSION}"
        "./bootstrap.sh"
        "./configure" (or "./configure --enable-usb")
        "make"
        <edit the configuration files>
        <test>
        "make install"
        <start the daemon>

Requirements:
The only known requirement to build thelinkbox is the GNU GCC C compiler, the
GNU GPP C++ compiler or something compatible, a make program and a Bourne 
shell to run the configure script. GNU make is not required, any old make 
should be fine. I tried to avoid the use of exotic compiler features, 
hopefully any version of GCC will work.  

If for some reason either configure or the make should fail please
send me the details and I'll try to help you correct the problem.

Extract:
The distribution tar file extracts into a version specific subdirectory so 
different releases do not conflict with each other.  Substitute the actual
version number for "{VERSION}" above.  The ".tgz" extension signifies a
Gzip'ed TAR file.

GNU configure:
Thelinkbox uses a GNU autoconfig generated configuration script to generate a 
site-specific Makefile that is then used to build thelinkbox.  The configuration
script is written to be as portable as possible by only assuming the 
availability of the most generic Bourne shell features.  A suitable shell 
should be available on just about every *nix system.  It will certainly be 
available on any system using other GNU tools.

The configuration script provides a great deal of flexibility in the way
the target program is built and installed, run "./configure --help" for the
gory details.  Luckily most of the features are seldom needed and running
configure without any options is usually sufficient.  

If you want to use an USB audio dongle for an interface specify the 
--enable-usb option when you run configure.  In this case will need to have 
the libusb and it's header files installed on your system.

If you have the GNU readline library (libreadline) you can configure to tlbcmd 
to use it by specifing the --with-readline switch.  The readline library
adds features such as command history and curses based line editing tlbcmd and
tlbchat.

Once the configure script has been run you should have a config.h and Makefile 
appropriate for your system.  

Make:
The generated makefile provides several useful targets, the default "all" 
builds thelinkbox and utilities.  The other frequently used targets include 
"install", "clean", "distclean", and "uninstall". 

There shouldn't be any errors or warnings displayed during the build process
for the core source files.  I've included several open source libraries with
thelinkbox to try to prevent "dependency hell", some of those libraries
generate warnings. If you get warnings or errors in the core source files 
I'd be interested in hearing about them.  

The final output of the build process is the thelinkbox daemon "tlb" which is 
built in the linkbox subdirectory.  We'll test it before installing it.  
Before we test it we need to edit the configuration files.

########################
thelinkbox Configuration
########################

A least two text files "tlb.conf" and "tlb.cmds" that are needed to configure 
thelinkbox.  The first file can contain all of the configuration necessary 
other than the mapping of DTMF digits to actions that are taken when the 
DTMF digits are received. The second file contains the mapping of DTMF digits 
to commands.

The port configuration can be included in the main file if desired, but
it's probably cleaner and less confusing to have one configuration file
per port.  The port configuration files are then included from the main
configuration file by use of the "include" directive.

The files files "tlb.conf.sample" and "tlb.cmds.sample" are provided as 
a starting point for your configuration files.  Since the sample files
include documentation on features that are not mentioned elsewhere it is 
worthwhile to review the sample configuration file with each release to 
discover new capabilities which you may want to take advantage of.  

There are sample port configuration files for compatible interfaces included
in the release, just make N copies of that file that matches the type
of interface you will be using and them modify them as needed.

Fire up your favorite editor and change the various variables to appropriate 
values.  Refer to the comments in the file for guidance.  

--- snip ---
% cd /usr/local/etc
% cp tlb.conf.sample tlb.conf
% cp tlb.cmds.sample tlb.cmds
% cp iMic.conf.sample Port1.conf
% vi tlb.conf
% vi tlb.cmds
% vi Port1.conf
% cp irlp.conf.sample Port2.conf
% vi Port2.conf
% cp cm108.conf.sample Port3.conf
% vi Port3.conf
--- snip ---

Lines beginning with ';' or '#' are comments, if you decide to set any of the
optional settings be sure to delete leading ';' character before the 
configuration variable.

Since daemons are not connected to any console they must communicate with 
the sysop in some other manner.  As with most *nix daemons thelinkbox can use 
the syslog system.  Thelinkbox uses "LOG_LOCAL5" as the facility when opening 
the log and generates messages with priorities LOG_INFO, LOG_WARNING and 
LOG_ERR.  It's up to the user to decide which message if any he wants to log,
but unless you are a psychic I would strongly suggest that at least the LOG_ERR 
messages be logged.
 
Thelink box can also generate a local log file without using the syslog system.
This has many advantages and is the recommended method.  The configuration
file variable LogFileRolloverType controls the logging method.

##################################
Determining physical USB Addresses
##################################

If you are not using multiple USB audio dongles you can safely ignore this
section.

Since most USB audio dongles do not have a unique serial numbers we must 
resort to configuring thelinkbox using physical USB addresses.  This means 
the physical port on the PC or USB hub that the USB audio dongle is plugged
into determines the configuration address. Unfortunately it is not 
easy to determine this address without more documentation than is usually 
available for PCs.  Even if documentation were available determining the 
correct address would be tedious and error prone. 

Luckily there's an easier way!

1. Configure thelinkbox for a single USB port.  Do no specify either
AudioDongleSN or AudioDongleAdr.

2. Connect a compatible dongle to ONE port you wish to use.  It's easy for 
the linkbox to find the port to use when there's only one dongle by searching 
all USB buses until a device is found.
  
3. Start thelinkbox in debug mode.  Hopefully thelinkbox will find the dongle
and display it's address.  Make a note of the address displayed. A sticky 
label next to the port might be a good idea.

--- snip ---
[root@localhost ~]# /root/tlb -d -f /home/skip/tbd/tbd.conf.440
thelinkbox Version 0.13 compiled Dec 14 2007 16:24:36
16:40:07 thelinkbox V 0.13 compiled Dec 14 2007 16:24:36 initializing
16:40:07 EndPointInit: Initializing 440 port (0)
16:40:07 EndPointInit: called, TxKeyMethod: 5, RxCosMethod: 6
16:40:07 UsbInit: found USB device @ 1-1.2
16:40:07 UsbInit: USB audio device for 440 is /dev/dsp2
16:40:07 UsbInit: Changing audio device from auto to /dev/dsp2
16:40:07 FindDevByAdr: event device for "1-1.2" is /dev/input/event3
16:40:07 Fragsize: 256, fragstotal: 5, bytes: 1280
16:40:07 EndPointInit: returning 0
--- snip ---

In this example the AudioDongleAdr that was discovered is 1-1.2.

4. Hit control C to shutdown thelinkbox.

5. Disconnect the dongle from the port just found and move it to the next
port.  Repeat steps 3 to 5 until the addresses for all of the ports that will
be used have been determined.

6. Configure thelinkbox for multiple ports specifying the addresses found
for AudioDongleAdr.  


#######
Testing
#######

The daemon has two command line switches to aid testing.  The first switch -f
specifies where the tlb.conf configuration is located.  

The second switch -d enables debug mode, causing the daemon to run in the
foreground as a user process while displaying debug information on the screen.
The debug switch may be used multiple times to increase the detail level of
the information displayed up to a maximum of three times.

For our purposes a single -d suffices. 

--- snip ---
[root@localhost ~]# /root/tlb -d -f /home/skip/tbd/tbd.conf.440
thelinkbox Version 0.33 compiled May 14 2008 16:31:19
6:45:49 thelinkbox V 0.33 compiled May 14 2008 16:31:19 initializing
6:45:49 EndPointInit: Initializing 440 port
6:45:49 EndPointInit: called, TxKeyMethod: 5, RxCosMethod: 6
6:45:49 UsbInit: found USB device "440" @ 1-2.2
6:45:49 UsbInit: USB audio device for 440 is /dev/dsp2
6:45:49 UsbInit: Changing audio device from auto to /dev/dsp2
6:45:49 FindInputDevBySn: event device for "440" is /dev/input/event3
6:45:49 AudioInit#1586: opening "/dev/dsp2"
6:45:49 Fragsize: 256, fragstotal: 5, bytes: 1280
6:45:49 EndPointInit: returning 0
GenAVRS(): sending string:
)EL-wb6ymh!3345.88NE11820.19W0PHG8660/449220/131 On  @0645
PullerLoginAck(): Client 8 successfully updated status.
6:45:50 Msg from EchoLink: EchoLink Server v2.5.996
6:45:50 Msg from EchoLink:
6:45:50 Msg from EchoLink: ECHO4: Scottsdale, AZ USA
Full station list downloaded successfully, 8166 stations listed.
--- snip ---

The message beginning with "PullerLoginAck():" indicates that the daemon 
was successful in logging into the EchoLink directory server.

The last message is the real good news, our directory request returned 7904
stations.  If your callsign or password were not recognized no stations would
be listed and you would see an error message from EchoLink simular to the
following:

--- snip ---
6:47:01 Msg from EchoLink: INCORRECT PASSWORD
6:47:01 Msg from EchoLink:
6:47:01 Msg from EchoLink: Please check the password
6:47:01 Msg from EchoLink: and try again. If you have
6:47:01 Msg from EchoLink: forgotten it, see the Validation
6:47:01 Msg from EchoLink: section at www.echolink.org.
Full station list downloaded successfully, 0 stations listed.
--- snip ---

If you are not able to get a station list, check your callsign and password
in the configuration file.  You might also want to rerun the test using more 
"-d" switches to help determine what when wrong.
        
NB: thelinkbox makes NO attempt to control the sound card's mixer.
You'll need to set levels manually with the mixer utility of your choice.

To test run thelinkbox in debug mode "linkbox/tlb -d -f tlb.conf". Log
messages will be displayed on the console. Key a radio and hopefully
you'll see "COS detected". Press some touchtone buttons and hopefully
you'll also see them displayed.

--- snip ---
[root@localhost ~]# /root/tlb -d -f /home/skip/tbd/tbd.conf.440
thelinkbox Version 0.13 compiled Dec 14 2007 16:24:36
16:48:41 thelinkbox V 0.13 compiled Dec 14 2007 16:24:36 initializing
16:48:41 EndPointInit: Initializing 440 port (0)
16:48:41 EndPointInit: called, TxKeyMethod: 5, RxCosMethod: 6
16:48:41 UsbInit: found USB device @ 1-1.2
16:48:41 UsbInit: USB audio device for 440 is /dev/dsp2
16:48:41 UsbInit: Changing audio device from auto to /dev/dsp2
16:48:41 FindDevByAdr: event device for "1-1.2" is /dev/input/event3
16:48:41 Fragsize: 256, fragstotal: 5, bytes: 1280
16:48:41 EndPointInit: returning 0
GenAVRS(): sending string:
)EL-wb6ymh!3345.88NE11820.19W0PHG8660/449220/131 On  @1648
PullerLoginAck(): Client 6 successfully updated status.
Full station list downloaded successfully, 7910 stations listed.
PollCOS: Node 440 COS active
PollCOS: Node 440 COS inactive
16:48:47 DecodeDTMF: Decoding "123"
16:48:47 RF user executing command "say I'm sorry dave I can't do that"
--- snip ---

[much more to come ... stay tuned!]

############
Installation
############

Thelinkbox is designed to run as a system daemon, i.e. a background program 
that's loaded automatically by the system as part of the bootup process. 

If you are not familiar with the system startup scripts or you are not
comfortable starting thelinkbox automatically you can always start thelinkbox 
daemon manually when desired.  Starting thelinkbox automatically is primarily 
needed when the host is unattended and it is desired to run thelinkbox 24/7.
Refer to the "Running without Root access" section if you would rather not
modify your system's startup behavior.

Unfortunately my installation rules and knowledge are not complete enough about
all of the various *nix variations to complete the installation without manual
assistance.

Most Posix operating systems start system daemons using approaches similar
to either FreeBSD (i.e. the BSD camp) or RedHat (i.e. the System V camp). 
Scripts to install thelinkbox on FreeBSD and RedHat Linux have been provided.
Start with the scripts that are the closest match to your system and then 
modify them if necessary.  I will be happy to include installation scripts 
for other operating system that are sent to me with future releases.

Configuration files for daemons are kept in different places on different
systems.  It doesn't really matter where the configuration file is as long

ChangeLog
#####################################
Changes in version 0.56 2/06/2021  wd5m

Bug Fixes:
1. Update nodes.cpp to call GetTxAudio from ~Node() function to fix
a memory leak caused by intermittent unconsumed audio buffers when nodes disconnect.
2. Update VoipNode::DeleteTx in voip.cpp to consume all audio buffers
for disconnected node as part of the memory leak fix.
3. 06/2020 In src/conference.c, changed " int  bFoundCallsign = FALSE;" to
"int  bFoundCallsign = TRUE;" to adapt to a change in IRLP speakfreely.

Minor enhancements:
1. Added -q command line switch to the .port command to allow suppression
of logging of the command.  Useful for CGI scripts that poll constantly
using the .port command to update realtime web pages. Similar to .users.
2. Corrected tlb.config.sample to change ";ShowIPAddresses = 0" to
";LogIPAddresses = 0"

#####################################
Changes in version 0.55-2 7/07/2019  wd5m

Bug Fixes:
1. Update nodes.cpp to use "unsigned char" on DTMF2CharLookup[] table
so Debian 9 64 bit compiler is happy.

#####################################
Changes in version 0.55.2 1/27/2017  wd5m

Minor enhancements:
1. Added Debian install scripts.

#####################################
Changes in version 0.55.1 4/16/2015  wd5m

Minor enhancements:
1. Increased line length allowed in config file for longer tone definitions.

#####################################
Changes in version 0.55 8/31/2013

Bug Fixes:
1. Corrected audio issues on Big Endian boxes with wave file playback
   and tone generation.

#####################################
Changes in version 0.54 7/7/2013

Minor enhancements:
1. Added patches from Kristoff - ON1ARF for sysfs GPIO access for PT.
   (PTT only, no COS support)
2. Corrected callsign announcements using PCM files.

#####################################
Changes in version 0.53 12/9/2012

Minor enhancement:
1. Added new option '5' for LogFileRolloverType for monthly logs.

#####################################
Changes in version 0.52 11/29/2012 (Limited Beta)

Minor enhancements:
1. Added new USB device ID for URI (CM108AH) courtesy of Roy K6MF.

Bug Fixes:
1. Corrected bug in LoadFile which caused odd crashes and general bad behavior
   when the configuration variables HelpFile or InfoFile where used.
   
2. Corrected attempt to use parallel port to unkey transmitter after
   open failed.

3. Corrected a bug in the APRS-IS code that caused tlb to loop using 100% 
   of the CPU time when the APRS-IS server communications timed out.
   Thanks to Rob VK5TRM for reporting and help debugging this problem.

4. Corrected several memset() calls where the value (0) and the length
   were swapped.  These errors were caught by newer versions of GCC and
   were apparently benign.

5. Corrected a bug which caused "conferences" created by the Rtp_port
   configuration file variable to be deleted when the last user logs
   out.
   
#####################################
Changes in version 0.51 12/19/2011 (Limited Beta)

Minor enhancements:
1. Added ConfCmdEnable configuration variable to allow commands sent in 
   EchoLink text packets to be controlled separately from UDP audio and text
   "conferencing".  Previously commands were disabled when ConfEnable was 
   set to zero.  This is primarily of use with "Transcoding" conferences 
   (reflectors).

#####################################
Changes in version 0.50 12/13/2011 (Limited Beta)

Bug Fixes:
1. Corrected a segfault which occurred when bind failed (Bind2IP, SFBind2IP, 
   or RTPBind2IP configuration variables set).
   

#####################################
Changes in version 0.49 12/10/2011 (Limited Beta)

Minor enhancements:
1. Added ConfTextEnable configuration variable to allow UDP packet forwarding
   of audio packets and text "chat" packets (EchoLink) to be controlled 
   separately.  This is primarily of use with "Transcoding" conferences 
   (reflectors).

#####################################
Changes in version 0.48 11/7/2011 (Limited Beta)

Minor enhancements:
1. Added -q command line switch to the .users command to allow suppression
   of logging of the command.  Useful for CGI scripts that poll constantly
   using the .users command to update realtime web pages.

Bug Fixes:
1. Corrected a bug which prevented audio from being sent to the DTMF decoder
   from VoIP ports when no physical ports are configured.
   
Misc changes:
1. Removed deprecated support for the chan_rtpdir interface to Asterisk which
   has been replaced with G.726.
   

#####################################
Changes in version 0.47 11/17/2010 (Limited Beta)

New features:
1. Added ability to create an RTP conference separate from the speak freely/RTP
conference previously provided.  This allows you to create connections from
any pair of ports, not just the ports specified by SF_Port and SF_ReplyPort.
This potentially allows multiple nodes to share the same IP address under
certain circumstances.  See the new RTP_Port and RTPBind2IP configuration 
variables for more information.

2. Added support for the G.726 codec.  The G.726 codec can be used with RTP
connections only. The G.726 codec is selected by using the -7 switch with
the .connect command.  This is the prefered codec for use between tlb nodes
and AllStar nodes.  Thanks to the assistance of Jim WB6NIL in designing and
debugging support for this codec.

Minor enhancements:
1. Added COS and PTT support for an PCF8754 i2c port expander chip connected
to an iMic.  The PCF8754 is available in a 16 pin DIP package which is
DIY friendly.  Additionally the i2c bus of the iMic is connected to a
8 pin SOIC EEPROM which has HUGE pins in comparison to the iMic itself.
Software replaces soldering skills, news at 11:00.

2. Added new command .txpower to set transmitter power.  Currently only
supported by the Doug Hall protocol.

3. Added default support for the CTCSS input on the URI courtesy of Jim WB6NIL.

4. Changed the amount of audio sent in an uLaw packet in RTP mode to 320 

src/users.c
/* Copyright (C) 2002 GFRN systems

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
   See the GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.

   The latest version of this program may be found at
   http://CQiNet.sourceforge.net

   $Log: users.c,v $
   Revision 1.8  2008/02/26 17:34:44  wb6ymh
   Added dmalloc support.

   Revision 1.7  2007/12/01 01:02:21  wb6ymh
   Deleted STRCMP macro, it's now in common.h.

   Revision 1.6  2007/06/27 19:21:15  wb6ymh
   Added code to DeleteUser to free the QTH string.

   Revision 1.5  2004/05/29 17:32:49  wb6ymh
   1. Eliminated unused and unneeded fields from the UserInfo structure to reduce
      memory requirements.
   2. Added delete of the user structure itself in DeleteUser() (!).

   Revision 1.4  2003/08/31 23:24:07  wb6ymh
   Corrected portability problem caused by the assumption that u_long was the
   same size as an IP address.

   Revision 1.3  2003/04/30 22:01:25  wb6ymh
   Modified LoadACL() and SaveACL() to use the AppName configuration variable's
   value as the application name when creating pid files rather than a
   hardcoded "tbd".

   Revision 1.2  2002/11/02 18:31:10  wb6ymh
   Added routines to load, save, modify and refresh the access control list.

   Revision 1.1.1.1  2002/08/10 20:33:41  wb6ymh
   initial import

*/

#include "common.h"

#ifndef _WIN32
   // FreeBSD, Linux, etc..
   #include <stdio.h>
   #include <stdlib.h>
   #ifdef TIME_WITH_SYS_TIME
      #include <sys/time.h>
      #include <time.h>
   #else
      #ifdef HAVE_SYS_TIME_H
         #include <sys/time.h>
      #else
         #include <time.h>
      #endif
   #endif
   #include <sys/socket.h>
   #include <netinet/in.h>
   #include <arpa/inet.h>
   #include <string.h>
   #include <ctype.h>
   #include <netdb.h>
   #include <sys/stat.h>
#else
   // Windoze
   #include <stdio.h>
   #include <stdlib.h>
   #include <string.h>
   #include <time.h>
   #include <winsock2.h>
   #include <ctype.h>
   #include <sys/stat.h>
#endif

#include "avl.h"
#include "main.h"
#include "users.h"
#include "configvars.h"
#include "hostfile.h"

#ifdef USE_DMALLOC
#include "dmalloc.h"
#endif

#define ACL_FILENAME_EXT   "acl"

struct avl_table *UserTree;
struct avl_table *UserIPAdrTree;
struct avl_table *ACL_Call_Tree;
struct avl_table *ACL_IP_Tree;

UserInfo *pUserDebug = NULL;

int Id = 1001;

void DeleteUser(UserInfo *pUser,int Virgin)
{
   D2PRINTF(("Deleting user \"%s\".\n",pUser->Callsign));
   if(!Virgin && avl_find(UserIPAdrTree,pUser) != NULL) {
      avl_delete(UserIPAdrTree,pUser);
   }

   if(!Virgin) {
      if(avl_delete(UserTree,pUser) == NULL) {
         LOG_ERROR(("DeleteUser(): avl_delete() failed to find user \"%s\".\n",
                    pUser->Callsign));
      }
   }

   if(pUser->Callsign != NULL) {
      free(pUser->Callsign);
   }

   if(pUser->Qth != NULL) {
       free(pUser->Qth);
   }

   free(pUser);
}

// User compares by callsign
int UserCompare(const void *avl_a,const void *avl_b,void *avl_param)
{
   UserInfo *p_a = (UserInfo *) avl_a;
   UserInfo *p_b = (UserInfo *) avl_b;

   return STRCMP(p_a->Callsign,p_b->Callsign);
}

// User compares by callsign
int UserIPCompare(const void *avl_a,const void *avl_b,void *avl_param)
{
   UserInfo *p_a = (UserInfo *) avl_a;
   UserInfo *p_b = (UserInfo *) avl_b;

   if(p_a->HisAdr.ADDR == p_b->HisAdr.ADDR) {
      return 0;
   }

   if(p_a->HisAdr.ADDR > p_b->HisAdr.ADDR) {
      return 1;
   }
   else {
      return -1;
   }
}

int ACL_CallCompare(const void *avl_a,const void *avl_b,void *avl_param)
{
   ACL_User *p_a = (ACL_User *) avl_a;
   ACL_User *p_b = (ACL_User *) avl_b;

   return STRCMP(p_a->Callsign,p_b->Callsign);
}

// User compares by callsign
int ACL_IPCompare(const void *avl_a,const void *avl_b,void *avl_param)
{
   ACL_User *p_a = (ACL_User *) avl_a;
   ACL_User *p_b = (ACL_User *) avl_b;

   if(p_a->HisAdr.ADDR == p_b->HisAdr.ADDR) {
      return 0;
   }

   if(p_a->HisAdr.ADDR > p_b->HisAdr.ADDR) {
      return 1;
   }
   else {
      return -1;
   }
}


UserInfo *CreateNewUser(char *Callsign)
{
   UserInfo *pUser = malloc(sizeof(UserInfo));

   if(pUser != NULL) {
      memset(pUser,0,sizeof(UserInfo));
      pUser->Callsign = strdup(Callsign);
      pUser->NodeID = Id++;
   }
   return pUser;
}

void DeleteACLUser(ACL_User *pACL)
{
   avl_delete(ACL_IP_Tree,pACL);
   avl_delete(ACL_Call_Tree,pACL);

   if(pACL->Callsign != NULL) {
      free(pACL->Callsign);
   }

   if(pACL->CallPlus != NULL) {
      free(pACL->CallPlus);
   }

   if(pACL->HostName != NULL) {
      free(pACL->HostName);
   }

   if(pACL->Password != NULL) {
      free(pACL->Password);
   }

   free(pACL);
}

// The following fields must be filled in: 
//    Callsign
//    HostName ("-" for no hostname checking) 
//    Password ("-" for no password checking)
//    bAuthorized
//    Optionally:
//    CallPlus - User's name (appended to Callsign to create CallPlus)
int ACLAddUser(ACL_User *pACL)
{
   int Ret = TRUE;
   ACL_User *pOldACL = NULL;
   ACL_User *pNewACL;
   struct hostent *pTemp;
   char *cp;
   int StrLen;

   pACL->HisAdr.ADDR = INADDR_NONE;

   if(*pACL->HostName != '-') {
   // A host name was specified, make sure it's valid
      pTemp = GetHostByName(pACL->HostName);
      if(pTemp == NULL) {
         Ret = FALSE;
      }
      else {
         pACL->HisAdr.ADDR = IP_FROM_HOSTENT(pTemp,0);
      }
   }

   if(Ret) {
   // The ACL Callsign field does not contain a "-L" or "-R", but the
   // callplus field does...
      if((cp = strchr(pACL->Callsign,'-')) != NULL) {
      // Remove -R or -L from the callsign
         *cp = 0;
      }

      if((pOldACL = avl_find(ACL_Call_Tree,pACL)) != NULL) {
      // delete old entry
         DeleteACLUser(pOldACL);
      }

      pNewACL = (ACL_User *) malloc(sizeof(ACL_User));
      if(pNewACL != NULL) {
         memset(pNewACL,0,sizeof(ACL_User));
         pNewACL->bRefreshed = pACL->bRefreshed;
         pNewACL->HostName = strdup(pACL->HostName);
         pNewACL->Callsign = strdup(pACL->Callsign);
         if(cp != NULL) {
            *cp = '-';
         }
         
         if(pACL->CallPlus != NULL) {
            StrLen = strlen(pACL->Callsign) + strlen(pACL->CallPlus) + 2;
            pNewACL->CallPlus = malloc(StrLen);
            if(pNewACL->CallPlus != NULL) {
               strcpy(pNewACL->CallPlus,pACL->Callsign);
               if(*pACL->CallPlus != '-') {
                  strcat(pNewACL->CallPlus," ");
               }
               strcat(pNewACL->CallPlus,pACL->CallPlus);
            }
         }
         else {
            pNewACL->CallPlus = strdup(pACL->Callsign);
         }
         
         if(pACL->Password != NULL) {
            pNewACL->Password = strdup(pACL->Password);
         }
         pNewACL->HisAdr.ADDR = pACL->HisAdr.ADDR;
         pNewACL->bAuthorized = pACL->bAuthorized;
         pNewACL->LastResolved = pACL->LastResolved;
         avl_insert(ACL_Call_Tree,pNewACL);
         if(*pNewACL->HostName != '-') {
            avl_insert(ACL_IP_Tree,pNewACL);
         }
      }
   }

   return Ret;
}

void RefreshACL()
{
   ACL_User *pACL = NULL;
   struct avl_traverser avl_trans;
   struct hostent *pTemp;

   pACL = (ACL_User *) avl_t_first(&avl_trans,ACL_Call_Tree);
   while(pACL != NULL) {
      if(*pACL->HostName != '-') {
         pTemp = GetHostByName(pACL->HostName);
         if(pTemp != NULL) {
            pACL->HisAdr.ADDR = IP_FROM_HOSTENT(pTemp,0);
         }
      }
      pACL = (ACL_User *) avl_t_next(&avl_trans);
   }
}

void LoadACL()
{
   char Line[80];
   char *Allow;
   ACL_User *pACL = NULL;
   ACL_User *pNextACL = NULL;
   struct avl_traverser avl_trans;
   char  *WhiteSpace = "\t\n ";
   int LineNum = 0;
   FILE *fp = NULL;
   ACL_User ACL_New;
   struct stat FileStats;
   char  AclFilename[80];

   snprintf(AclFilename,sizeof(AclFilename),"%s." ACL_FILENAME_EXT,AppName);
   
   if((fp = fopen(AclFilename,"r")) == NULL) {
      if(stat(AclFilename,&FileStats) == 0) {
      // The ACL file exists but we couldn't open it, complain
         LOG_ERROR(("LoadACL(): Unable to open ACL file.\n"));
      }
      return;
   }

   memset(&ACL_New,0,sizeof(ACL_New));
   ACL_New.bRefreshed = TRUE;
   while(fgets(Line,sizeof(Line),fp) != NULL) {
      LineNum++;

      Allow = strtok(Line,WhiteSpace);
      if(Allow == NULL || *Allow == '#' || *Allow == ';') {
      // Empty line or comment
         continue;
      }

      if(STRCMP(Allow,"allow") == 0) {
         ACL_New.bAuthorized = TRUE;
      }
      else if(STRCMP(Allow,"deny") == 0) {
         ACL_New.bAuthorized = FALSE;
      }
      else {
      // Invalid line
         LOG_WARN(("LoadACL(): Ignoring line %d, invalid action \"%s\".\n",
                   LineNum,Allow));
         continue;
      }

      ACL_New.Callsign = strtok(NULL,WhiteSpace);
      if(ACL_New.Callsign == NULL) {
      // Invalid line
         LOG_WARN(("LoadACL(): Ignoring line %d, callsign missing.\n",LineNum));
         continue;
      }

      if((ACL_New.HostName = strtok(NULL,WhiteSpace)) == NULL) {
         LOG_WARN(("LoadACL(): Ignoring line %d, hostname missing.\n",LineNum));
         continue;
      }

      if((ACL_New.Password = strtok(NULL,WhiteSpace)) == NULL) {
         LOG_WARN(("LoadACL(): Ignoring line %d, password missing.\n",LineNum));
         continue;
      }
      
      ACL_New.CallPlus = strtok(NULL,"\n");

   // Skip any leading whitespace

      if(ACL_New.CallPlus != NULL) {
         while(*ACL_New.CallPlus && isspace(*ACL_New.CallPlus)) {
            ACL_New.CallPlus++;
         }
      }

      if(!ACLAddUser(&ACL_New)) {
         LOG_WARN(("LoadACL(): Ignoring line %d, unable to resolve "
                   "hostname \"%s\".\n",LineNum,ACL_New.HostName));
      }
   }
   fclose(fp);

// Remove any old entries from tree
   
   pNextACL = (ACL_User *) avl_t_first(&avl_trans,ACL_Call_Tree);
   while((pACL = pNextACL) != NULL) {
   // NB: Get the next user now in case we deleted the current user
      pNextACL = (ACL_User *) avl_t_next(&avl_trans);
      if(!pACL->bRefreshed) {
         DeleteACLUser(pACL);
      }
      pACL->bRefreshed = FALSE;
   }
}

// ACL format:
// <"allow"|"deny>"<tab><callsign><tab><hostname/IP Adr><tab><password><tab>
// <Name/Location>
//
// The callsign field is required and should be entered in upper case without 
// any suffix (-R or -L).  Stations are allowed/denied based on the base 
// callsign.
//
// hostname can be a numeric IP address, actual resolvable hostname or "-" for 
// no hostname
//
// The password can be "-" for no password or the actual password. The password
// may not contain spaces or tabs.
//
// The Name/location field is free text and extends to the end of the line.
// If it is desired to display <callsign>-R or <callsign>-L on the Echolink
// station list start the name/location field with "-R" or "-L".
//
void SaveACL()
{
   struct avl_traverser avl_trans;
   ACL_User *pACL;
   char *cp;
   FILE *fp = NULL;
   char  AclFilename[80];

   snprintf(AclFilename,sizeof(AclFilename),"%s." ACL_FILENAME_EXT,AppName);
   if((fp = fopen(AclFilename,"w")) == NULL) {
      LOG_ERROR(("SaveACL(): Unable to open ACL file for write.\n"));
      return;
   }

   pACL = (ACL_User *) avl_t_first(&avl_trans,ACL_Call_Tree);
   while(pACL != NULL) {
      if(pACL->CallPlus == NULL) {
         cp = NULL;
      }
      else if((cp = strchr(pACL->CallPlus,'-')) == NULL) {
         cp = strchr(pACL->CallPlus,' ');
         if(cp != NULL) {
         // Skip the blank
            cp++;
         }
      }
      
      fprintf(fp,"%s\t%s\t%s\t%s\t%s\n",
              pACL->bAuthorized ? "allow" : "deny",
              pACL->Callsign,
              pACL->HostName,
              pACL->Password,
              cp == NULL ? "" : cp);

      pACL = (ACL_User *) avl_t_next(&avl_trans);
   }
   fclose(fp);
}


